### PR TITLE
Pickle support for NT

### LINKED ISSUE #104198: NestedTensor cannot be pickled
```python
import torch
import pickle

a, b = torch.arange(3), torch.arange(5) + 3
nt = torch.nested.nested_tensor([a, b])
with open('tensor.pkl', 'wb') as f:
    pickle.dump(nt, f)
```

<details>

<summary>
click to see original report
</summary>

### 🚀 The feature, motivation and pitch

Hello,

I am trying to make a custom dataset class of ZINC dataset on PyTorch geometric with nested tensors using "torch.nested.nested_tensor", and when executing I got the following error: 

"Internal error: NestedTensorImpl doesn't support sizes."

Is there a way to get the of the different tensors in the nested tensor? 

Thanks

### Alternatives

_No response_

### Additional context

_No response_

</details>

cc @mruberry @mikaylagawarecki @cpuhrsch @jbschlosser @bhosmer @drisspg

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110219

Fixes #104198

### CODE REVIEW

File: tools/autograd/gen_python_functions.py [hunk 1]
```diff
@@ -159,7 +159,6 @@
     "fill.Scalar",  # only used by the functionalization pass
     "lift.*",
     "normal_functional",  # only used by the functionalization pas
-    "_nested_view_from_buffer",  # View only version of _nested_from_buffer. This will force users to only use the "safe" version.
```
Comment (albanD) [inline tools/autograd/gen_python_functions.py:162]: What about using the "safe" version? :p

File: torch/_tensor.py [hunk 2]
```diff
@@ -366,6 +366,14 @@ def _reduce_ex_internal(self, proto):
                 ),
             )
             return (torch._utils._rebuild_sparse_tensor, args_sparse_compressed)
+        elif self.is_nested:
+            args_nested = (
+                self.values(),
```
Comment (cpuhrsch) [inline torch/_tensor.py:371]: This won't work for arbitrary NestedTensors though right?

Comment (jbschlosser) [reply torch/_tensor.py:371]: Ok true, `values()` has a contiguity restriction :(

File: test/test_nestedtensor.py [hunk 3]
```diff
@@ -2917,6 +2918,15 @@ def grad_test_func(a, b, c):
 
         gradcheck(grad_test_func, inputs=(a, b, c), check_batched_grad=False)
 
+    @dtypes(torch.float, torch.double, torch.half)
+    @parametrize("requires_grad", [False, True])
+    def test_pickle(self, device, dtype, requires_grad):
+        a = random_nt_from_dims([3, None, 5], device=device, dtype=dtype,
+                                requires_grad=requires_grad)
+        serialized = pickle.dumps(a)
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_nestedtensor.py:2926]: good point, fixed

File: tools/autograd/gen_python_functions.py (hunk 1 above)
Comment (jbschlosser) [reply tools/autograd/gen_python_functions.py:162]: I'm not entirely clear on this "safe" version comment (cc @drisspg). AFAIK this is the only way to construct an NT directly from its parts without copying, and `_nested_from_buffer` doesn't exist despite what the comment says :p

Imo we should either expose this to python or provide a proper public API for this functionality, and the latter would take design work that isn't relevant to our current focus: the subclass version of NT.

**Review by cpuhrsch — APPROVED** [review]

File: torch/_tensor.py (hunk 2 above)
Comment (jbschlosser) [reply torch/_tensor.py:374]: Discussed offline: eventually introduce a private API that does what `values()` does now and deprecate `values()` in its current form, to be replaced by appropriate usage of `view(-1)` / `reshape(-1)`.

File: tools/autograd/gen_python_functions.py (hunk 1 above)
Comment (jbschlosser) [reply tools/autograd/gen_python_functions.py:162]: Discussed offline with Driss: we're okay opening this private API up for now

**Pull request closed without merging**